File: android/OpenStream100Remote/app/src/main/java/org/openstream100/remote/RemoteApi.kt
package org.openstream100.remote

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import android.net.Uri
import org.json.JSONObject
import java.io.ByteArrayOutputStream
import java.net.HttpURLConnection
import java.net.URL
import java.util.UUID

const val PROTOCOL_VERSION = 1

data class RemoteSettings(
    val server: String,
    val token: String,
    val fingerprint: String = "",
)

data class MixerChannel(
    val index: Int,
    val label: String,
    val colour: String,
    val available: Boolean,
    val muted: Boolean,
    val level: Float,
    val meterLeft: Float,
    val meterRight: Float,
    val iconPath: String?,
)

data class MixerAction(
    val index: Int,
    val id: String,
    val label: String,
)

data class MixerState(
    val revision: Long,
    val page: Int,
    val pageCount: Int,
    val channels: List<MixerChannel>,
    val actions: List<MixerAction>,
)

class RemoteApi(settings: RemoteSettings) {
    private val baseUrl = normaliseServer(settings.server)
    private val token = settings.token.trim()

    suspend fun state(): MixerState = withContext(Dispatchers.IO) {
        val payload = request("GET", "/state")
        if (payload.optInt("protocol") != PROTOCOL_VERSION) {
            error("The computer uses an unsupported remote protocol.")
        }
        val channelsJson = payload.getJSONArray("channels")
        val channels = buildList {
            for (index in 0 until channelsJson.length()) {
                val item = channelsJson.getJSONObject(index)
                add(
                    MixerChannel(
                        index = item.getInt("index"),
                        label = item.optString("label", "Disabled"),
                        colour = item.optString("color", "#5B82F6"),
                        available = item.optBoolean("available"),
                        muted = item.optBoolean("muted"),
                        level = item.optDouble("level").toFloat().coerceIn(0f, 1f),
                        meterLeft = item.optDouble("meter_left").toFloat().coerceIn(0f, 1f),
                        meterRight = item.optDouble("meter_right").toFloat().coerceIn(0f, 1f),
                        iconPath = item.optString("icon").takeIf { it.isNotBlank() },
                    )
                )
            }
        }
        val actionsJson = payload.getJSONArray("actions")
        val actions = buildList {
            for (index in 0 until actionsJson.length()) {
                val item = actionsJson.getJSONObject(index)
                add(
                    MixerAction(
                        index = item.getInt("index"),
                        id = item.optString("id", "disabled"),
                        label = item.optString("label", "Disabled"),
                    )
                )
            }
        }
        MixerState(
            revision = payload.optLong("revision"),
            page = payload.optInt("page"),
            pageCount = payload.optInt("page_count", 1).coerceAtLeast(1),
            channels = channels,
            actions = actions,
        )
    }

    suspend fun setVolume(page: Int, channel: Int, level: Float) {
        command("set_volume", page, channel, level.coerceIn(0f, 1f))
    }

    suspend fun toggleMute(page: Int, channel: Int) {
        command("toggle_mute", page, channel)
    }

    suspend fun selectPage(page: Int) {
        command("select_page", page)
    }

    suspend fun pressButton(page: Int, button: Int) {
        command("press_button", page, button)
    }

    suspend fun icon(path: String): ByteArray = withContext(Dispatchers.IO) {
        requestBytes("GET", path)
    }

    private suspend fun command(
        name: String,
        page: Int,
        channel: Int? = null,
        value: Float? = null,
    ) = withContext(Dispatchers.IO) {
        val payload = JSONObject()
            .put("protocol", PROTOCOL_VERSION)
            .put("request_id", UUID.randomUUID().toString())
            .put("command", name)
            .put("page", page)
        if (channel != null) payload.put("channel", channel)
        if (value != null) payload.put("value", value.toDouble())
        request("POST", "/command", payload)
    }

    private fun request(method: String, path: String, body: JSONObject? = null): JSONObject {
        return JSONObject(requestBytes(method, path, body).toString(Charsets.UTF_8))
    }

    private fun requestBytes(
        method: String,
        path: String,
        body: JSONObject? = null,
    ): ByteArray {
        val endpoint = if (path.startsWith("/api/")) {
            "$baseUrl$path"
        } else {
            "$baseUrl/api/v1$path"
        }
        val connection = URL(endpoint).openConnection() as HttpURLConnection
        try {
            connection.requestMethod = method
            connection.connectTimeout = 1_500
            connection.readTimeout = 1_500
            connection.setRequestProperty("Accept", "application/json")
            connection.setRequestProperty("Authorization", "Bearer $token")
            if (body != null) {
                connection.doOutput = true
                connection.setRequestProperty("Content-Type", "application/json")
                connection.outputStream.use { it.write(body.toString().toByteArray()) }
            }
            val status = connection.responseCode
            val stream = if (status in 200..299) connection.inputStream else connection.errorStream
            val bytes = stream?.use { input ->
                ByteArrayOutputStream().use { output ->
                    input.copyTo(output)
                    output.toByteArray()
                }
            } ?: byteArrayOf()
            if (status !in 200..299) {
                val message = runCatching {
                    JSONObject(bytes.toString(Charsets.UTF_8)).optString("error")
                }.getOrNull()
                error(message?.takeIf { it.isNotBlank() } ?: "Computer returned HTTP $status.")
            }
            return bytes
        } finally {
            connection.disconnect()
        }
    }

    companion object {
        suspend fun requestPinPairing(
            server: String,
            deviceId: String,
            deviceName: String,
        ): Int = withContext(Dispatchers.IO) {
            val baseUrl = normaliseServer(server)
            val payload = JSONObject()
                .put("protocol", PROTOCOL_VERSION)
                .put("device_id", deviceId)
                .put("device_name", deviceName)
            val connection = URL("$baseUrl/api/v1/pair/request")
                .openConnection() as HttpURLConnection
            try {
                connection.requestMethod = "POST"
                connection.connectTimeout = 2_000
                connection.readTimeout = 2_000
                connection.doOutput = true
                connection.setRequestProperty("Accept", "application/json")
                connection.setRequestProperty("Content-Type", "application/json")
                connection.outputStream.use { output ->
                    output.write(payload.toString().toByteArray())
                }
                val status = connection.responseCode
                val stream = if (status in 200..299) {
                    connection.inputStream
                } else {
                    connection.errorStream
                }
                val response = stream?.bufferedReader()?.use { it.readText() }.orEmpty()
                val document = runCatching { JSONObject(response) }.getOrElse {
                    error("The computer returned an invalid pairing response.")
                }
                if (status !in 200..299) {
                    error(document.optString("error").ifBlank { "Could not request a PIN." })
                }
                document.optJSONObject("pairing")?.optInt("expires_in", 120) ?: 120
            } finally {
                connection.disconnect()
            }
        }

        suspend fun pairWithPin(
            server: String,
            pin: String,
            deviceId: String,
            deviceName: String,
        ): RemoteSettings = withContext(Dispatchers.IO) {
            val cleanPin = pin.filter(Char::isDigit)
            require(cleanPin.length == 6) { "Enter the six-digit PIN shown on the computer." }
            val baseUrl = normaliseServer(server)
            val payload = JSONObject()
                .put("protocol", PROTOCOL_VERSION)
                .put("pin", cleanPin)
                .put("device_id", deviceId)
                .put("device_name", deviceName)
            val connection = URL("$baseUrl/api/v1/pair/complete")
                .openConnection() as HttpURLConnection
            try {
                connection.requestMethod = "POST"
                connection.connectTimeout = 2_000
                connection.readTimeout = 2_000
                connection.doOutput = true
                connection.setRequestProperty("Accept", "application/json")
                connection.setRequestProperty("Content-Type", "application/json")
                connection.outputStream.use { output ->
                    output.write(payload.toString().toByteArray())
                }
                val status = connection.responseCode
                val stream = if (status in 200..299) {
                    connection.inputStream
                } else {
                    connection.errorStream
                }
                val response = stream?.bufferedReader()?.use { it.readText() }.orEmpty()
                val document = runCatching { JSONObject(response) }.getOrElse {
                    error("The computer returned an invalid pairing response.")
                }
                if (status !in 200..299) {
                    error(document.optString("error").ifBlank { "Pairing failed." })
                }
                val token = document.optString("token")
                require(token.length >= 32) { "The computer did not return a valid device token." }
                RemoteSettings(
                    server = baseUrl,
                    token = token,
                    fingerprint = document.optString("token_fingerprint"),
                )
            } finally {
                connection.disconnect()
            }
        }

        fun normaliseServer(value: String): String {
            val trimmed = value.trim().trimEnd('/')
            require(trimmed.isNotBlank()) { "Enter the computer address." }
            return if (trimmed.startsWith("http://") || trimmed.startsWith("https://")) {
                trimmed
            } else {
                "http://$trimmed"
            }
        }

        fun parsePairingUri(value: String): RemoteSettings {
            val uri = Uri.parse(value.trim())
            require(uri.scheme == "openstream100" && uri.host == "pair") {
                "This is not an OpenStream100 pairing code."
            }
            require(uri.getQueryParameter("protocol") == PROTOCOL_VERSION.toString()) {
                "The pairing code uses an unsupported protocol."
            }
            val server = uri.getQueryParameter("server").orEmpty()
            val token = uri.getQueryParameter("token").orEmpty()
            val fingerprint = uri.getQueryParameter("fingerprint").orEmpty()
            require(server.isNotBlank() && token.length >= 32) {
                "The pairing code is incomplete."
            }
            return RemoteSettings(server, token, fingerprint)
        }
    }
}

File: ios/OpenStream100Remote/OpenStream100Remote/RemoteAPI.swift
import Foundation

actor RemoteAPI {
    private let baseURL: URL
    private let token: String
    private let session: URLSession

    init(settings: RemoteSettings, session: URLSession = .shared) throws {
        baseURL = try Self.normalizedServer(settings.server)
        token = settings.token.trimmingCharacters(in: .whitespacesAndNewlines)
        self.session = session
    }

    func state() async throws -> MixerState {
        let data = try await request(method: "GET", path: "/state")
        let state = try JSONDecoder().decode(MixerState.self, from: data)
        guard state.protocolVersion == remoteProtocolVersion else {
            throw RemoteError.unsupportedProtocol
        }
        return state
    }

    func setVolume(page: Int, channel: Int, level: Double) async throws {
        try await command("set_volume", page: page, channel: channel, value: min(max(level, 0), 1))
    }

    func toggleMute(page: Int, channel: Int) async throws {
        try await command("toggle_mute", page: page, channel: channel)
    }

    func selectPage(_ page: Int) async throws {
        try await command("select_page", page: page)
    }

    func pressButton(page: Int, button: Int) async throws {
        try await command("press_button", page: page, channel: button)
    }

    func icon(path: String) async throws -> Data {
        try await request(method: "GET", path: path)
    }

    private func command(
        _ name: String,
        page: Int,
        channel: Int? = nil,
        value: Double? = nil
    ) async throws {
        var body: [String: Any] = [
            "protocol": remoteProtocolVersion,
            "request_id": UUID().uuidString,
            "command": name,
            "page": page,
        ]
        if let channel { body["channel"] = channel }
        if let value { body["value"] = value }
        _ = try await request(method: "POST", path: "/command", json: body)
    }

    private func request(method: String, path: String, json: [String: Any]? = nil) async throws -> Data {
        let url: URL
        if path.hasPrefix("/api/") {
            guard let resolved = URL(string: path, relativeTo: baseURL)?.absoluteURL else {
                throw RemoteError.invalidAddress
            }
            url = resolved
        } else {
            url = baseURL.appending(path: "api/v1\(path)")
        }

        var request = URLRequest(url: url, timeoutInterval: 1.5)
        request.httpMethod = method
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        if let json {
            request.httpBody = try JSONSerialization.data(withJSONObject: json)
            request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        }
        let (data, response) = try await session.data(for: request)
        try Self.validate(data: data, response: response)
        return data
    }

    static func requestPINPairing(server: String, deviceID: String, deviceName: String) async throws -> Int {
        let baseURL = try normalizedServer(server)
        let body: [String: Any] = [
            "protocol": remoteProtocolVersion,
            "device_id": deviceID,
            "device_name": deviceName,
        ]
        let data = try await unauthenticatedPOST(baseURL: baseURL, path: "api/v1/pair/request", body: body)
        guard
            let object = try JSONSerialization.jsonObject(with: data) as? [String: Any],
            let pairing = object["pairing"] as? [String: Any]
        else { return 120 }
        return pairing["expires_in"] as? Int ?? 120
    }

    static func pairWithPIN(
        server: String,
        pin: String,
        deviceID: String,
        deviceName: String
    ) async throws -> RemoteSettings {
        let cleanPIN = pin.filter(\.isNumber)
        guard cleanPIN.count == 6 else { throw RemoteError.invalidPIN }
        let baseURL = try normalizedServer(server)
        let body: [String: Any] = [
            "protocol": remoteProtocolVersion,
            "pin": cleanPIN,
            "device_id": deviceID,
            "device_name": deviceName,
        ]
        let data = try await unauthenticatedPOST(baseURL: baseURL, path: "api/v1/pair/complete", body: body)
        guard let object = try JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            throw RemoteError.invalidResponse
        }
        let token = object["token"] as? String ?? ""
        guard token.count >= 32 else { throw RemoteError.invalidResponse }
        return RemoteSettings(
            server: baseURL.absoluteString.trimmingCharacters(in: CharacterSet(charactersIn: "/")),
            token: token,
            fingerprint: object["token_fingerprint"] as? String ?? ""
        )
    }

    static func normalizedServer(_ value: String) throws -> URL {
        var trimmed = value.trimmingCharacters(in: .whitespacesAndNewlines)
        while trimmed.hasSuffix("/") { trimmed.removeLast() }
        guard !trimmed.isEmpty else { throw RemoteError.invalidAddress }
        if !trimmed.hasPrefix("http://") && !trimmed.hasPrefix("https://") {
            trimmed = "http://\(trimmed)"
        }
        guard let components = URLComponents(string: trimmed),
              let scheme = components.scheme,
              ["http", "https"].contains(scheme),
              components.host != nil,
              let url = components.url
        else { throw RemoteError.invalidAddress }
        return url
    }

    static func parsePairingURI(_ value: String) throws -> RemoteSettings {
        guard let components = URLComponents(string: value.trimmingCharacters(in: .whitespacesAndNewlines)),
              components.scheme == "openstream100",
              components.host == "pair"
        else { throw RemoteError.invalidPairingCode }
        let parameters = Dictionary(uniqueKeysWithValues: (components.queryItems ?? []).map { ($0.name, $0.value ?? "") })
        guard parameters["protocol"] == String(remoteProtocolVersion) else {
            throw RemoteError.unsupportedProtocol
        }
        let server = parameters["server"] ?? ""
        let token = parameters["token"] ?? ""
        guard !server.isEmpty, token.count >= 32 else { throw RemoteError.incompletePairingCode }
        return RemoteSettings(server: server, token: token, fingerprint: parameters["fingerprint"] ?? "")
    }

    private static func unauthenticatedPOST(baseURL: URL, path: String, body: [String: Any]) async throws -> Data {
        let url = baseURL.appending(path: path)
        var request = URLRequest(url: url, timeoutInterval: 2)
        request.httpMethod = "POST"
        request.httpBody = try JSONSerialization.data(withJSONObject: body)
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        let (data, response) = try await URLSession.shared.data(for: request)
        try validate(data: data, response: response)
        return data
    }

    private static func validate(data: Data, response: URLResponse) throws {
        guard let http = response as? HTTPURLResponse else { throw RemoteError.invalidResponse }
        guard (200...299).contains(http.statusCode) else {
            let object = try? JSONSerialization.jsonObject(with: data) as? [String: Any]
            let message = object?["error"] as? String
            throw RemoteError.server(message?.isEmpty == false ? message! : "Computer returned HTTP \(http.statusCode).")
        }
    }
}
